### ISSUE TITLE
Problem using switch!() with exhaustive patterns

### ISSUE THREAD

**Issue by jugglerchris** [Geal/nom#316]
As part of #271, I found that the following:

``` rust
switch!(value!(some_bool),
    true => ...,
    false => ...)
```

gives an "unreachable pattern" error because `true`/`false` are exhaustive, but the macro inserts a default `_ => ...` case too.  https://github.com/rust-lang/rust/issues/4112 suggests that adding an `if true` guard is a workaround, but that isn't accepted by the macro.

My workaround was to wrap the `bool` in an otherwise useless `Option`:

``` rust
switch!(value!(Some(some_bool)),
    Some(true) => ...,
    Some(false) => ...)
```

**Comment by Geal** [comment]
That's an annoying bug. I could make a different version of switch, but I may be able to detect the _ too. I'll have to think about it.

**Comment by porglezomp** [comment]
Not a solution for everyone, but if anyone else is running into this because they're trying to write a catch-all pattern, I found that instead of doing
```rust
switch!(u8,
  0x02 => something!() |
  x => something_with!(x),
)
```
you can get away with doing
```rust
switch!(u8,
  0x02 => something!() |
  x@0x00...0xFF => something_with!(x),
)
```

**Comment by Geal** [comment]
Fixed in #466
